lib/SPIRV/SPIRVReader.cpp

```diff
@@ -381,11 +390,13 @@ Type *SPIRVToLLVM::transType(SPIRVType *T, bool IsClassMember) {
   }
   case internal::OpTypeTokenINTEL:
     return mapType(T, Type::getTokenTy(*Context));
-  case OpTypePointer:
-    return mapType(
-        T, PointerType::get(
-               transType(T->getPointerElementType(), IsClassMember),
-               SPIRSPIRVAddrSpaceMap::rmap(T->getPointerStorageClass())));
+  case OpTypePointer: {
+    unsigned AS = SPIRSPIRVAddrSpaceMap::rmap(T->getPointerStorageClass());
+    Type *ElementTy = transType(T->getPointerElementType(), UseTPT);
+    if (UseTPT)
+      return TypedPointerType::get(ElementTy, AS);
```

Comment:
Why can't we use `transType` result as-is in this case?

Reply:
This is part of the opaque pointers work. When reading SPIR-V IR, we need to keep track of what the typed pointer equivalent of all of the SPIR-V types are for eventual feeding into the name mangler. When `transType` normally returns a `PointerType` (whether special SPIR-V opaque type or a SPIR-V pointer type), we need to keep track elsewhere than the resulting LLVM IR type of what the type is.

Previously, this was done with `PointerIndirectPair`, but this patch replaces that representation with `TypedPointerType`. This also removes the duplication of `transType` just to compute `PointerIndirectPair`/`TypedPointerType`, but in doing so, it requires more care: we don't want to cache a result of `TypedPointerType` in the type map (since `TypedPointerType` cannot be used in LLVM IR), and if we want to compute a `TypedPointerType`, we can't reuse the cached `PointerType` in the map, since it doesn't have the info we need.

As for worries about type duplication: LLVM uniques most types already, with struct types being the case where it's possible to get type duplication. However, most of the struct types go through the `getOrCreateOpaqueStructType` helper, which will ensure that there's no duplication of names. The ones that don't go through `OpTypeStruct`, where there is an explicit check to reuse the type map, even if `UseTPT` is true (as struct types don't pass `UseTPT` into recursive type construction--we never create structs whose elements may be `TypedPointerType`, as we don't need them for mangling purposes).

I can clarify this in source code comments for the next revision.

Reply:
Thanks for the explanation, I now get it: in some code paths we at first produce actual `typed` pointers to generate proper mangling and then we opaquify them in order to emit correct LLVM IR